### ISSUE TITLE
`F.grid_sample` fails to dispatch correctly when args are of different subclasses

### ISSUE THREAD

**Issue by jph00** [pytorch/pytorch#47091]
## 🐛 Bug

`F.grid_sample` fails to dispatch correctly when args are of different subclasses on PyTorch 1.7 on Ubuntu 20.04.

This works correctly:

```
import torch
t1 = tensor([1.]).view(1,1,1,1)
t2 = tensor([1.,1.]).view(1,1,1,2)
F.grid_sample(t1, t2)
```

    >>> tensor([[[[0.2500]]]])

However this raises "no implementation found for 'torch.nn.functional.grid_sample'":

```
import torch
class _T1(torch.Tensor): ...
class _T2(torch.Tensor): ...
t1 = tensor([1.]).view(1,1,1,1).as_subclass(_T1)
t2 = tensor([1.,1.]).view(1,1,1,2).as_subclass(_T2)
F.grid_sample(t1, t2)
```

I don't see any reason why dispatch should fail in this case - it looks like a bug to me. However, I might be missing something.

Sorry @mruberry for the mass of bug reports! Trying to get fastai working on pytorch 1.7 so I'm letting you know what I find as I go along...

cc @hameerabbasi @rgommers

**Comment by mruberry** [comment]
@jph00 No apologies needed, these are great! I think we would accept a PR fixing this.

**Comment by hameerabbasi** [comment]
This one is actually according to the RFC, so it's more of a design decision as to whether the RFC should be changed. Specifically, in [this section](https://github.com/pytorch/rfcs/blob/master/RFC-0001-torch-function-for-methods.md#proposal)

> This has the side-effect of ensuring unrelated class trees are not merged, which is an inconsistency in NumPy's own design. Specifically, consider the example of two direct subclasses of torch.Tensor. Both will return NotImplemented, and therefore, the check will fail and TypeError will be raised.

By "unrelated class trees", I mean that not all the classes have a common descendant that we can return the type of. You wouldn't want the output type to depend on the order of the input types, only the types themselves.

**Comment by jph00** [comment]
Thanks for the explanation @hameerabbasi . Unfortunately, it does mean that subclassing Tensor causes breakage in things that work fine with Tensor itself, as in this case. I currently use this very hacky workaround in our base class for all fastai tensors (`first` is the same as `next`, but returns `None` instead of raising `StopIteration`):

```python
class TensorBase(Tensor):
    def __torch_function__(self, func, types, args=(), kwargs=None):
        # workaround for https://github.com/pytorch/pytorch/issues/47091
        if len(types)>1:
            t = first(o for o in types if issubclass(o,TensorBase))
            if t: types = tuple(t if issubclass(o, TensorBase) else o for o in types)
        return super().__torch_function__(func, types, args=args, kwargs=kwargs)
```

Is there some more elegant and correct way I could work around this issue?

**Comment by hameerabbasi** [comment]
> Is there some more elegant and correct way I could work around this issue?

I would suggest something like the following, so it affects *only your subclasses*. Incidentally, if you're doing this kind of thing it might be wise to wrap `Tensor` instead of subclass it, but that ship has sailed, I think. Anyways here's a small sampler of what can be done.

```python
import torch
from torch import _C

class MyTensor(torch.Tensor):
    @classmethod
    def __torch_function__(cls, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}

        with _C.DisableTorchFunction():
            ret = func(*args, **kwargs)
            return _convert(ret, cls)

def _convert(ret, cls):
    if isinstance(ret, torch.Tensor):
        ret = ret.as_subclass(cls)

    if isinstance(ret, (tuple, list)):
        ret = type(ret)(_convert(r, cls) for r in ret)

    return ret

class MyTensor2(torch.Tensor):
    pass

if __name__ == "__main__":
    a = torch.tensor([5]).as_subclass(MyTensor)
    b = torch.tensor([6]).as_subclass(MyTensor2)

    print(type(a + b))  # MyTensor
    print(type(b + a))  # MyTensor
```

**Comment by jph00** [comment]
i'm happy to go the wrapping approach, although iirc i'd been warned off it before as being hard to get full compatibility

i've tried this

```
class TensorBase:
    def __init__(self, data, metadata=None, **kwargs):
        self._fa_tensor = tensor(data)
        for k,v in kwargs.items(): setattr(self, k, v)
    def __torch_function__(self, func, types, args=(), kwargs=None):
        if kwargs is None: kwargs = {}
        args = [getattr(a,'_fa_tensor',a) for a in args]
        ret = func(*args, **kwargs)
        return TensorBase(ret, **self.__dict__)
    def __getattr__(self, k): return getattr(self._fa_tensor, k)
    def __getitem__(self, k): return self._fa_tensor[k]
```

but i'm not sure how to support stuff like + with an int, without doing something a bit hacky

**Comment by hameerabbasi** [comment]
> But I'm not sure how to support stuff like `+` with an int, without doing something a bit hacky

There's an example `Wrapper` class here that's more or less tested (via `gradcheck`): https://github.com/pytorch/pytorch/blob/master/test/test_overrides.py#L659

**Comment by jph00** [comment]
Thanks for the link. Unfortunately that fails a lot of our tests, e.g.:

- `len`
- pickling
- indexing with a tensor mask

It feels more fragile to me than subclassing, and seems likely to be a higher maintenance burden. What do you think? What's the benefit you see vs subclassing?

**Comment by hameerabbasi** [comment]
> What do you think? What's the benefit you see vs subclassing?

Mainly the fact that one subclass can only inherit the properties of e.g. `Tensor`, whereas with wrapping, you're using composition, so you don't have to worry about that. Of course, this is Python, so one can easily use metaclasses or construct classes at runtime that inherit from the subclass you like AND some base class.

**Comment by jph00** [comment]
Thanks - because we use a single base class for all fastai tensors, which includes all functionality needed by child classes, we don't have a problem there, so I guess we'll stick with subclassing.

I've tested your suggest fix using `torch._C` and it works great. Many thanks!

**Comment by hameerabbasi** [comment]
@jph00 @ezyang Is it okay if we close this one as `wontfix`?

**Comment by ezyang** [comment]
Yup

**Comment by jph00** [comment]
@ezyang @hameerabbasi After a few months of experience with this, I've decided that the current behavior is actually quite nice. It just means we have to be more careful about typing everything, and we have to have classes "opt in" to participating in particular operations. This is how we do that:

https://github.com/fastai/fastai/blob/master/fastai/torch_core.py#L370

If you have any suggestions for better ways to do this, please let us know. :)
